### ISSUE TITLE
Investigate smarter diffing algorithm

### ISSUE THREAD

**Issue by fitzgen** [fitzgen/dodrio#28]
Some initial profiling shows that we spend about

* 10% of time in rendering,
* 5% of time diffing to generate a change list, and
* 75% of time applying a change list diff

Our current diffing algorithm is a rather simple, greedy, single-pass algorithm. It seems we could probably spend a little bit more time in diffing if it generated a smaller change list and then we would likely see some perf wins.

**Comment by JakubKoralewski** [comment]
I'm no expert here, but wouldn't Svelte's *[Virtual Dom is pure overhead](https://svelte.dev/blog/virtual-dom-is-pure-overhead)* approach be more efficient than a VDOM? Is it compatible with the current architecture?

**Comment by fitzgen** [comment]
I experimented with a similar approach and found it to yield better latency but worse throughput. That is often the right trade off. However, I did find that it was, in my experimental prototype, a more constrained paradigm and further work would be required in figuring out how to enable more freedom and developer ergonomics. I would encourage anyone who wanted to explore that paradigm in Rust to do so!

Anyways, I think this issue can be closed, since master is a bit smarter now.
